src/svg/helper/PatternManager.ts

```diff
@@ -0,0 +1,258 @@
+                imageSrc = pattern.image;
+            }
+            else if (pattern.image instanceof HTMLImageElement) {
+                imageSrc = pattern.image.src;
+            }
+            else if (pattern.image instanceof HTMLCanvasElement) {
+                imageSrc = pattern.image.toDataURL();
+            }
+
+            if (imageSrc === img.getAttribute('href')) {
+                // The same as old pattern image
+                return;
+            }
+
+            if (imageSrc) {
+                const x = pattern.x || 0;
+                const y = pattern.y || 0;
+                const rotation = pattern.rotation / Math.PI * 180;
+                const scaleX = pattern.scaleX || 1;
+                const scaleY = pattern.scaleY || 1;
+                const transform = `translate(${x}, ${y}) rotate(${rotation}) scale(${scaleX}, ${scaleY})`;
+                patternDom.setAttribute('patternTransform', transform);
+
+                img.setAttribute('href', imageSrc);
+                img.setAttribute('x', '0');
+                img.setAttribute('y', '0');
+
+                getPatternSize(pattern, (width, height) => {
+                    const w = width + '';
+                    const h = height + '';
+                    img.setAttribute('width', w);
+                    img.setAttribute('height', h);
+                    patternDom.setAttribute('width', w);
+                    patternDom.setAttribute('height', h);
+                });
+
+                patternDom.appendChild(img);
+                patternDomMap.set(pattern, patternDom);
+            }
+        }
+    }
+
+    /**
+     * Mark a single pattern to be used
+     *
+     * @param displayable displayable element
+     */
+    markUsed(displayable: Displayable) {
+        if (displayable.style) {
+            if (isPattern(displayable.style.fill)) {
+                super.markDomUsed(patternDomMap.get(displayable.style.fill));
+            }
+            if (isPattern(displayable.style.stroke)) {
+                super.markDomUsed(patternDomMap.get(displayable.style.fill));
+            }
+        }
+    }
+
+}
+
+
+const patternSizeCache = new LRU<CachedImageObj>(50);
+
+type CachedImageObj = {
+    width: number,
+    height: number
+};
+
+function getPatternSize(
+    pattern: PatternObject,
+    cb?: (width: number, height: number) => void
+) {
+    if (pattern.image instanceof HTMLImageElement) {
+        if (typeof cb === 'function') {
+            cb(pattern.image.width, pattern.image.height);
+        }
+        return;
+    }
+
+    let src;
+    if (typeof pattern.image === 'string') {
+        src = pattern.image;
+        const size = patternSizeCache.get(src);
+        if (size) {
+            if (typeof cb === 'function') {
+                cb(size.width, size.height);
+            }
+            return;
+        }
+    }
+    else if (pattern.image instanceof HTMLCanvasElement) {
+        src = pattern.image.toDataURL();
+    }
```

Comment:
Can use canvas.width, canvas.height directly here. It's just a remind since `getPatternSize` can be removed and replaced with `createOrUpdateImage`.